File: dynamic_array_test.go
package dynamic_array

import (
	"testing"
)

func TestDynamicArray(t *testing.T) {
	new_array := Array{}
	t.Run("Create array tests", func(t *testing.T) {

		t.Run("Expect creation of a new array to return true", func(t *testing.T) {
			if new_array.Create(5) != true {
				t.Fail()
			}

		})

		t.Run("Expect created array to have the specified capacity", func(t *testing.T) {
			if new_array.Capacity != 5 {
				t.Fail()
			}
		})

		t.Run("Expect created array to have a length of 0", func(t *testing.T) {
			if new_array.Length != 0 {
				t.Fail()
			}
		})

	})

	t.Run("IsOutOfBounds tests", func(t *testing.T) {
		t.Run("Expect IsOutOfBounds func to return true if index exheed length", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.IsOutOfBounds(4) != true {
				t.Fail()
			}
		})

		t.Run("Expect IsOutOfBounds func to return true if index is less than 0", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.IsOutOfBounds(-1) != true {
				t.Fail()
			}
		})
	})

	t.Run("Add to array tests", func(t *testing.T) {

		t.Run("Expect add function to return true if the operation was successful", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			if new_array.Add(10) != true {
				t.Fail()
			}
		})

		t.Run("Expect length to increase by one each time we add new element", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3)
			if new_array.Length != 1 {
				t.Fail()
			}
			new_array.Add(7)
			if new_array.Length != 2 {
				t.Fail()
			}
			new_array.Add(8)
			if new_array.Length != 3 {
				t.Fail()
			}
		})

		t.Run("Expect array to contain added element", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3)
			if new_array.Contains(3) != true {
				t.Fail()
			}
			new_array.Add(7)
			if new_array.Contains(7) != true {
				t.Fail()
			}
			new_array.Add(8)
			if new_array.Contains(8) != true {
				t.Fail()
			}
		})

		t.Run("Expect capacity to be doubled if treshold is exheeded", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			new_array.Add(9) // index 3
			if new_array.Capacity != 6 {
				t.Fail()
			}
		})

	})

	t.Run("Remove from array tests", func(t *testing.T) {

		t.Run("Expect remove function to return element that was removed", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.RemoveAt(2) != 8 {
				t.Fail()
			}
		})

		t.Run("Expect length to decrease by one each time we remove an element", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3)  // index 0
			new_array.Add(7)  // index 1
			new_array.Add(8)  // index 2
			new_array.Add(9)  // index 3
			new_array.Add(10) // index 4
			new_array.RemoveAt(0)
			if new_array.Length != 4 {
				t.Fail()
			}
			new_array.RemoveAt(1)
			if new_array.Length != 3 {
				t.Fail()
			}
			new_array.RemoveAt(2)
			if new_array.Length != 2 {
				t.Fail()
			}
		})

		t.Run("Expect array not to contain the removed element after removal", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.RemoveAt(0)
			if new_array.Contains(3) != false {
				t.Fail()
			}
		})

		t.Run("Expect capacity to equal length after removal", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			new_array.Add(9) // index 3
			new_array.RemoveAt(3)
			if new_array.Length != 3 {
				t.Fail()
			}
			if new_array.Capacity != 3 {
				t.Fail()
			}
		})

	})

	t.Run("Size tests", func(t *testing.T) {

		t.Run("Expect size method to return length of the dynamic array", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.Size() != 3 {
				t.Fail()
			}
		})
	})

	t.Run("IsEmpty method tests", func(t *testing.T) {

		t.Run("Expect IsEmpty Method to return false if the array is not empty", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2

			if new_array.IsEmpty() != false {
				t.Fail()
			}
		})

		t.Run("Expect IsEmpty Method to return true if the array is empty", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			if new_array.IsEmpty() != true {
				t.Fail()
			}
		})
	})

	t.Run("Clear method tests", func(t *testing.T) {

		t.Run("Expect Clear method to return true", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.Clear() != true {
				t.Fail()
			}
		})

		t.Run("Expect Length to be 0", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			new_array.Clear()
			if new_array.Length != 0 {
				t.Fail()
			}
		})
	})

	t.Run("Get method tests", func(t *testing.T) {

		t.Run("Expect Get method to return element at index", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.Get(2) != 8 {
				t.Fail()
			}
		})

	})

	t.Run("Set method tests", func(t *testing.T) {

		t.Run("Expect Set method to return true is set was successful", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.Set(2, 10) != true {
				t.Fail()
			}
		})

		t.Run("Expect element at index to change to new element", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			new_array.Set(2, 10)
			if new_array.Inner_Array[2] != 10 {
				t.Fail()
			}
		})

	})

	t.Run("IndexOf method tests", func(t *testing.T) {

		t.Run("Expect IndexOf method to return -1 if element was not found", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.IndexOf(4) != -1 {
				t.Fail()
			}
		})

		t.Run("Expect IndexOf method to return element index if element was  found", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.IndexOf(8) != 2 {
				t.Fail()
			}
		})

		t.Run("Expect IndexOf method to return first index of element encountered in case when there is a duplicate ", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			new_array.Add(8) // index 3
			if new_array.IndexOf(8) != 2 {
				t.Fail()
			}
		})

	})

	t.Run("LastIndexof method tests", func(t *testing.T) {

		t.Run("Expect LastIndexof method to return -1 if element was not found", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.IndexOf(4) != -1 {
				t.Fail()
			}
		})

		t.Run("Expect LastIndexOf method to return last index of element encountered in case when there is a duplicate ", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			new_array.Add(8) // index 3
			if new_array.LastIndexOf(8) != 3 {
				t.Fail()
			}
		})

	})

	t.Run("Contains method tests", func(t *testing.T) {

		t.Run("Expect Contains method to return false if element was not found", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.Contains(4) != false {
				t.Fail()
			}
		})

		t.Run("Expect Contains method to return true if element was not found", func(t *testing.T) {
			new_array := Array{}
			new_array.Create(3)
			new_array.Add(3) // index 0
			new_array.Add(7) // index 1
			new_array.Add(8) // index 2
			if new_array.Contains(3) != true {
				t.Fail()
			}
		})

	})

}

File: Array/dynamic_array.test.js
const Array = require("./dynamic_array")

describe('create method', () => {
    let new_array = new Array(5)

    test("Expect created array to have the specified capacity", function () {
        expect(new_array.capacity).toBe(5)
    })
    test("Expect created array to have a length of 0", function () {
        expect(new_array.length).toBe(0)
    })
})


describe('checkbounds method', () => {
    let new_array = new Array(5)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect checkbounds function to return true if index exheed length", function () {
        expect(() => new_array.checkBound(4)).toThrow("index is out of bound")
    })
    test("Expect  checkbounds  function to return true if index is less than 0", function () {
        expect(() => new_array.checkBound(-2)).toThrow("index is out of bound")
    })
    // test("Expect created array to have a length of 0", function () {
    //     expect(new_array.length).toBe(0)
    // })

})


describe('Add method', () => {
    let new_array = new Array(5)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Add method top return true", function () {
        expect(new_array.add(9)).toBe(true)
    })
    test("Expect length to increase by one each time we add new element", function () {
        let new_array = new Array(5)
        new_array.add(10)
        expect(new_array.length).toBe(1)
        new_array.add(11)
        expect(new_array.length).toBe(2)
        new_array.add(12)
        expect(new_array.length).toBe(3)
        new_array.add(13)
        expect(new_array.length).toBe(4)
    })

    test("Expect array to contain added element", function () {
        let new_array = new Array(5)
        new_array.add(10)
        expect(new_array.contains(10)).toBe(true)
        new_array.add(11)
        expect(new_array.contains(11)).toBe(true)
        new_array.add(12)
        expect(new_array.contains(12)).toBe(true)
        new_array.add(13)
        expect(new_array.contains(13)).toBe(true)
    })



    test("Expect capacity to be doubled if treshold is exheeded", function () {
        let new_array = new Array(3)
        new_array.add(10)
        new_array.add(11)
        new_array.add(12)
        new_array.add(13)
        expect(new_array.capacity).toBe(6)
    })

})






describe('Remove method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect remove function to return element that was removed", function () {
        expect(new_array.removeAt(1)).toBe(7)
    })
    test("Expect length to decrease by one each time we remove an element", function () {
        let new_array = new Array(5)
        new_array.add(3) // index 0
        new_array.add(7) // index 1
        new_array.add(8) // index 2
        new_array.add(11) // index 3
        new_array.removeAt(3)
        expect(new_array.length).toBe(3)
        new_array.removeAt(2)
        expect(new_array.length).toBe(2)
        new_array.removeAt(1)
        expect(new_array.length).toBe(1)
    })

    test("Expect array not to contain the removed element after removal", function () {
        let new_array = new Array(5)
        new_array.add(10)
        new_array.add(11)
        new_array.add(12)
        new_array.removeAt(0)
        expect(new_array.contains(10)).toBe(false)
    })



    test("Expect capacity to equal length after removal", function () {
        let new_array = new Array(3)
        new_array.add(10)
        new_array.add(11)
        new_array.add(12)
        new_array.add(13)
        new_array.removeAt(3)
        expect(new_array.length).toBe(3)
        expect(new_array.capacity).toBe(3)
    })

})





describe('size method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect size method to return length of the dynamic array", function () {
        expect(new_array.size()).toBe(3)
    })
})


describe('IsEmpty method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect IsEmpty Method to return false if the array is not empty", function () {
        expect(new_array.isEmpty()).toBe(false)
    })
    test("Expect IsEmpty Method to return true if the array is empty", function () {
        let new_array = new Array(3)
        expect(new_array.isEmpty()).toBe(true)
    })
})


describe('clear method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect Clear method to return true", function () {
        expect(new_array.clear()).toBe(true)
        expect(new_array.length).toBe(0)
    })

})



describe('Get method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect Get method to return element at index", function () {
        expect(new_array.get(2)).toBe(8)
    })

})


describe('set method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect Set method to return true is set was successful", function () {
        expect( new_array.set(2,10)).toBe(true)
    })
    test("Expect element at index to change to new element", function () {
        expect( new_array.array[2]).toBe(10)
    })
})


describe('indexOf method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect IndexOf method to return -1 if element was not found", function () {
        expect( new_array.indexOf(4)).toBe(-1)
    })
    test("Expect IndexOf method to return element index if element was  found", function () {
        expect( new_array.indexOf(8)).toBe(2)
    })
    test("Expect IndexOf method to return element index if element was  found", function () {
        new_array.add(8) // index 3
        expect( new_array.indexOf(8)).toBe(2)
    })
})



describe('contains method', () => {
    let new_array = new Array(3)
    new_array.add(3) // index 0
    new_array.add(7) // index 1
    new_array.add(8) // index 2
    test("Expect Contains method to return false if element was not found", function () {
        expect( new_array.contains(4)).toBe(false)
    })
    test("Expect Contains method to return true if element was not found", function () {
        expect( new_array.contains(3)).toBe(true)
    })
})